framework/src/media/MediaRecorderImpl.cpp

```diff
@@ -73,74 +85,111 @@ recorder_result_t MediaRecorderImpl::destroy()
+{
+	medvdbg("MediaRecorderImpl::destroyRecorder(recorder_result_t& ret)\n");
+
+	if (mCurState != RECORDER_STATE_IDLE) {
+		meddbg("MediaRecorderImpl::destroyRecorder(recorder_result_t& ret) - mCurState != RECORDER_STATE_IDLE\n");
+		return notifySync();
+	}
 
 	mCurState = RECORDER_STATE_NONE;
-	return RECORDER_OK;
+	ret = RECORDER_OK;
+	notifySync();
 }
 
 recorder_result_t MediaRecorderImpl::prepare()
 {
 	std::unique_lock<std::mutex> lock(mCmdMtx);
 
 	medvdbg("MediaRecorderImpl::prepare()\n");
-
-	if (mCurState != RECORDER_STATE_IDLE || mOutputDataSource == nullptr) {
-		meddbg("MediaRecorderImpl::prepare() - mCurState != RECORDER_STATE_IDLE || mOutputDataSource == nullptr\n");
+	RecorderWorker& mrw = RecorderWorker::getWorker();
+	if (!mrw.isAlive()) {
+		meddbg("MMediaRecorderImpl::prepare()- (mrw.isAlive() == false)\n");
 		return RECORDER_ERROR;
 	}
-
-	RecorderWorker& mrw = RecorderWorker::getWorker();
-	mrw.getQueue().enQueue(&MediaRecorderImpl::notifySync, shared_from_this());
+	recorder_result_t ret = RECORDER_ERROR;
+	mrw.getQueue().enQueue(&MediaRecorderImpl::prepareRecorder, shared_from_this(), std::ref(ret));
 	mSyncCv.wait(lock);
 
-	if (set_audio_stream_in(mOutputDataSource->getChannels(), mOutputDataSource->getSampleRate(),
-							(pcm_format)mOutputDataSource->getPcmFormat()) != AUDIO_MANAGER_SUCCESS) {
+	return ret;
+}
 
-		meddbg("MediaRecorderImpl::prepare() - set_audio_stream_in() != AUDIO_MANAGER_SUCCESS\n");
-		return RECORDER_ERROR;
+void MediaRecorderImpl::prepareRecorder(recorder_result_t& ret)
+{
+	medvdbg("MediaRecorderImpl::prepareRecorder(recorder_result_t& ret)\n");
+
+	if (mCurState != RECORDER_STATE_IDLE || mOutputDataSource == nullptr) {
+		meddbg("MediaRecorderImpl::prepareRecorder(recorder_result_t& ret) - " \
+			"mCurState != RECORDER_STATE_IDLE || mOutputDataSource == nullptr\n");
+		return notifySync();
+	}
+
+	if (set_audio_stream_in(mOutputDataSource->getChannels(), mOutputDataSource->getSampleRate(),
+		(pcm_format)mOutputDataSource->getPcmFormat()) != AUDIO_MANAGER_SUCCESS) {
+		meddbg("MediaRecorderImpl::prepareRecorder(recorder_result_t& ret) - " \
+			"set_audio_stream_in(mOutputDataSource->getChannels(), mOutputDataSource->getSampleRate()," \
+			"(pcm_format)mOutputDataSource->getPcmFormat()) != AUDIO_MANAGER_SUCCESS\n");
+		return notifySync();
 	}
 
 	if (!mOutputDataSource->open()) {
-		meddbg("MediaRecorderImpl::prepare() - mOutputDataSource->open() failed\n");
-		return RECORDER_ERROR;
+		meddbg("MediaRecorderImpl::prepareRecorder(recorder_result_t& ret) - mOutputDataSource->open() failed\n");
+		return notifySync();
 	}
 
 	mBuffSize = get_input_frames_byte_size(get_input_frame_count());
 
 	if (mBuffSize < 0) {
-		meddbg("MediaRecorderImpl::prepare() - get_input_frames_byte_size(get_input_frame_count()) failed\n");
-		return RECORDER_ERROR;
+		meddbg("MediaRecorderImpl::prepareRecorder(recorder_result_t& ret) - (mBuffSize < 0)\n");
```

Comment:
1. if there's error case due to some variable's value condition, we'd better print out value to check error case properly
Also I think get_input_frame_count also checked in log.
2. you can see return value of get_input_frames_byte_size, it return 0 if pcm is null. so it should be changed to <=0

Reply:
1. if mBuffsize is 0 --> success case.
2. ok. i will add get_input_frame_count in log.

Reply:
yes i will fix that.